pkg/querier/batch/stream.go

```diff
@@ -39,64 +45,120 @@ func (bs *batchStream) atTime() int64 {
+			// then the one provided.
+			result = append(result, createBatch(valueTypes))
+		}
+		b = &result[resultLen-1]
+	}
+
+	ensureBatchType := func(valueTypes chunkenc.ValueType) {
+		for b.ValueTypes != valueTypes {
+			if b.ValueTypes == chunkenc.ValNone {
+				// Uninitialized Batch
+				initBatchType(b, valueTypes)
+				return
+			}
+			// These should be rare, means that we're merging into something where the chunk type changed
+			if b.Index == 0 {
+				// Batch not written yet
+				result[resultLen-1] = createBatch(valueTypes)
+				b = &result[resultLen-1]
+				return
+			}
+			// Batch already started, finish and start new
+			// We don't know its type so let the loop run one more time
+			nextBatch(valueTypes)
+		}
+	}
+
+	// Ensure that the batch we're writing to is not full and of the right type.
+	ensureBatch := func(valueTypes chunkenc.ValueType) {
 		if b.Index == size {
-			// The batch reached it intended size.
+			// The batch reached its intended size.
 			// Add another batch the the result
 			// and use it for further appending.
-
-			// The Index is the place at which new sample
-			// has to be appended, hence it tells the length.
-			b.Length = b.Index
-			resultLen++
-			if resultLen > len(result) {
-				// It is possible that result can grow longer
-				// then the one provided.
-				result = append(result, chunk.Batch{})
-			}
-			b = &result[resultLen-1]
+			nextBatch(valueTypes)
 		}
+		ensureBatchType(valueTypes)
 	}
 
-	for left.hasNext() && right.hasNext() {
-		checkForFullBatch()
-		t1, t2 := left.atTime(), right.atTime()
-		if t1 < t2 {
-			b.Timestamps[b.Index], b.Values[b.Index] = left.at()
-			left.next()
-		} else if t1 > t2 {
-			b.Timestamps[b.Index], b.Values[b.Index] = right.at()
-			right.next()
+	for {
+		if lt, rt := left.hasNext(), right.hasNext(); lt != chunkenc.ValNone && rt != chunkenc.ValNone {
+			t1, t2 := left.atTime(), right.atTime()
+			if t1 < t2 {
+				ensureBatch(lt)
+				populate(b, left, lt)
+				left.next()
+			} else if t1 > t2 {
+				ensureBatch(rt)
+				populate(b, right, rt)
+				right.next()
+			} else {
+				if (rt == chunkenc.ValHistogram || rt == chunkenc.ValFloatHistogram) && lt == chunkenc.ValFloat {
+					// Prefer historgrams over floats. Take left side if both has histograms.
+					ensureBatch(rt)
+					populate(b, right, rt)
+				} else {
+					ensureBatch(lt)
+					populate(b, left, lt)
+				}
+				left.next()
+				right.next()
+			}
+			b.Index++
 		} else {
-			b.Timestamps[b.Index], b.Values[b.Index] = left.at()
-			left.next()
-			right.next()
+			break
 		}
-		b.Index++
 	}
 
 	// This function adds all the samples from the provided
 	// batchStream into the result in the same order.
 	addToResult := func(bs batchStream) {
-		for ; bs.hasNext(); bs.next() {
-			checkForFullBatch()
-			b.Timestamps[b.Index], b.Values[b.Index] = bs.at()
-			b.Index++
-			b.Length++
+		for {
+			if t := bs.hasNext(); t != chunkenc.ValNone {
```

Comment:
Isn't easier this way?

```suggestion
		for t := bs.hasNext(); t != chunkenc.ValNone;  {
```

Reply:
reverted